### PR TITLE
Zipfldr

### PR DESCRIPTION
**Pull request by learn-more**
Initial implementation

JIRA issue: [CORE-7684](https://jira.reactos.org/browse/CORE-7684)

This pr aims to cleanup this code, and get it in the tree.
Further improvements will be made later on, like fixing `SHCreateFileExtractIconW`, implementing a toolbar that will handle `IExplorerCommand` etc. [CORE-14175](https://jira.reactos.org/browse/CORE-14175)

### CODE REVIEW

File: dll/shellext/zipfldr/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,61 @@
+
+set_cpp(WITH_RUNTIME WITH_EXCEPTIONS)
+if(NOT MSVC)
+    # HACK: this should be enabled globally!
+    add_compile_flags_language("-std=c++11" "CXX")
```
Comment (learn-more) [inline dll/shellext/zipfldr/CMakeLists.txt]: sorry thomas ;)

File: dll/shellext/zipfldr/precomp.h [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+#include <atlstr.h>
+#include <atlwin.h>
+#include <shlwapi.h>
+#include <shellapi.h>
+#include <strsafe.h>
+#include <rosdlgs.h>
+#include <assert.h>
+
+EXTERN_C
+ULONG DbgPrint(PCSTR Format,...);
```
Comment (learn-more) [inline dll/shellext/zipfldr/precomp.h]: We should think about moving this somewhere shared...

Comment (HBelusca) [reply dll/shellext/zipfldr/precomp.h]: Isn't it already in debug.h ?

File: dll/shellext/zipfldr/CMakeLists.txt (hunk 1 above)
Comment (sanchaez) [reply dll/shellext/zipfldr/CMakeLists.txt]: Isn't it? AFAIK rapps builds with this flag.

Comment (gigaherz) [reply dll/shellext/zipfldr/CMakeLists.txt]: This doesn't seem to indicate so: https://git.reactos.org/?p=reactos.git&a=search&h=HEAD&st=grep&s=c%2B%2B11

File: dll/shellext/zipfldr/precomp.h (hunk 2 above)
Comment (learn-more) [reply dll/shellext/zipfldr/precomp.h]: `boot/freeldr/freeldr/include/debug.h`
That doesn't really help outside freeldr ;)

Comment (HBelusca) [reply dll/shellext/zipfldr/precomp.h]: Ah, I thought it was also in [sdk/include/reactos/debug.h](https://git.reactos.org/?p=reactos.git;a=blob;f=sdk/include/reactos/debug.h;hb=851a554b5922af7ac1a21b12fd788ebb051641b8#l27).

File: modules/rostests/apitests/shell32/IShellFolderViewCB.cpp [hunk 3]
```diff
@@ -0,0 +1,622 @@
...
+#define SFVM_GETEXTVIEWS              40 /* undocumented */
+#define SFVM_GET_CUSTOMVIEWINFO       77 /* undocumented */
+#define SFVM_ENUMERATEDITEMS          79 /* undocumented */
+#define SFVM_GET_VIEW_DATA            80 /* undocumented */
+#define SFVM_GET_WEBVIEW_LAYOUT       82 /* undocumented */
+#define SFVM_GET_WEBVIEW_CONTENT      83 /* undocumented */
+#define SFVM_GET_WEBVIEW_TASKS        84 /* undocumented */
+#define SFVM_GET_WEBVIEW_THEME        86 /* undocumented */
+#define SFVM_GETDEFERREDVIEWSETTINGS  92 /* undocumented */
+#endif
```
Comment (HBelusca) [inline modules/rostests/apitests/shell32/IShellFolderViewCB.cpp]: Using some sdk/include/reactos/undocshell.h for these would be nice.

File: dll/shellext/zipfldr/CExplorerCommand.cpp [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+    }
+
+    // *** IEnumExplorerCommand methods ***
+    STDMETHODIMP Next(ULONG celt, IExplorerCommand **pUICommand, ULONG *pceltFetched)
+    {
+        if (!pUICommand)
+            return E_POINTER;
+
+        if (pceltFetched)
+            *pceltFetched = 0;
```
Comment (HBelusca) [inline dll/shellext/zipfldr/CExplorerCommand.cpp]: Is it some kind of "FALSE"?

File: dll/shellext/zipfldr/CZipEnumerator.hpp [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+ * PURPOSE:     CZipEnumerator
+ * COPYRIGHT:   Copyright 2017 Mark Jansen (mark.jansen@reactos.org)
+ */
+
+struct CZipEnumerator
+{
+private:
+    CComPtr<IZip> spZip;
+    bool first;
+    CAtlList<CStringA> _returned;
```
Comment (HBelusca) [inline dll/shellext/zipfldr/CZipEnumerator.hpp]: You don't use the usual "m_Blah;" notation for class members here?

File: dll/shellext/zipfldr/CZipExtract.cpp [hunk 6]
```diff
@@ -0,0 +1,477 @@
...
+                }
+            }
+
+            do
+            {
+                err = unzReadCurrentFile(uf, Buffer, sizeof(Buffer));
+
+                if (err < 0)
+                {
+                    __debugbreak();
```
Comment (HBelusca) [inline dll/shellext/zipfldr/CZipExtract.cpp]: Are these debugbreaks meant to be removed in the near future?

File: dll/shellext/zipfldr/CZipFolder.hpp [hunk 7]
```diff
@@ -0,0 +1,627 @@
...
+    DWORD dwDefaultState;
+    int cxChar;
+    int fmt;
+};
+
+static FolderViewColumns g_ColumnDefs[] = 
+{
+    { IDS_COL_NAME,     SHCOLSTATE_TYPE_STR | SHCOLSTATE_ONBYDEFAULT,   25, LVCFMT_LEFT },
+    { IDS_COL_TYPE,     SHCOLSTATE_TYPE_STR | SHCOLSTATE_ONBYDEFAULT,   20, LVCFMT_LEFT },
+    { IDS_COL_COMPRSIZE,SHCOLSTATE_TYPE_STR | SHCOLSTATE_ONBYDEFAULT,   10, LVCFMT_RIGHT },
```
Comment (HBelusca) [inline dll/shellext/zipfldr/CZipFolder.hpp]: /me suggests space after the first comma.

File: dll/shellext/zipfldr/CZipFolder.hpp [hunk 8]
```diff
@@ -0,0 +1,627 @@
...
+        bool isDir = zipEntry->ZipType == ZIP_PIDL_DIRECTORY;
+        switch (iColumn)
+        {
+        case 0: /* Name, ReactOS specific? */
+            return GetDisplayNameOf(pidl, 0, &psd->str);
+        case 1: /* Type */
+        {
+            SHFILEINFOA shfi;
+            DWORD dwAttributes = isDir ? FILE_ATTRIBUTE_DIRECTORY : FILE_ATTRIBUTE_NORMAL;
+            ULONG_PTR firet = SHGetFileInfoA(zipEntry->Name, dwAttributes, &shfi, sizeof(shfi), SHGFI_USEFILEATTRIBUTES | SHGFI_TYPENAME);
```
Comment (HBelusca) [inline dll/shellext/zipfldr/CZipFolder.hpp]: "firet"?

File: dll/shellext/zipfldr/CZipFolder.hpp [hunk 9]
```diff
@@ -0,0 +1,627 @@
...
+            else if (child2->mkid.cb)
+                result = -1;
+        }
+
+        return MAKE_COMPARE_HRESULT(result);
+    }
+    STDMETHODIMP CreateViewObject(HWND hwndOwner, REFIID riid, LPVOID *ppvOut)
+    {
+        static const GUID UnknownIID = // {93F81976-6A0D-42C3-94DD-AA258A155470}
+        {0x93F81976, 0x6A0D, 0x42C3, {0x94, 0xDD, 0xAA, 0x25, 0x8A, 0x15, 0x54, 0x70}};
```
Comment (HBelusca) [inline dll/shellext/zipfldr/CZipFolder.hpp]: There's no existing standard name for this one?

File: dll/shellext/zipfldr/CZipExtract.cpp (hunk 6 above)
Comment (learn-more) [reply dll/shellext/zipfldr/CZipExtract.cpp]: yeah, I probably should replace them with DPRINT1 before committing.

File: dll/shellext/zipfldr/CZipFolder.hpp (hunk 8 above)
Comment (learn-more) [reply dll/shellext/zipfldr/CZipFolder.hpp]: `FileInfo RETurn` I guess, probably copied that from somwhere.

File: dll/shellext/zipfldr/CZipFolder.hpp (hunk 9 above)
Comment (learn-more) [reply dll/shellext/zipfldr/CZipFolder.hpp]: Couldn't find it.

File: dll/shellext/zipfldr/CZipFolder.hpp (hunk 8 above)
Comment (gigaherz) [reply dll/shellext/zipfldr/CZipFolder.hpp]: FileInfo-RET

File: dll/shellext/zipfldr/CZipFolder.hpp (hunk 9 above)
Comment (gigaherz) [reply dll/shellext/zipfldr/CZipFolder.hpp]: Apparently not. I did a quick googling and no one has a name for it.

File: dll/shellext/zipfldr/CExplorerCommand.cpp (hunk 4 above)
Comment (learn-more) [reply dll/shellext/zipfldr/CExplorerCommand.cpp]: 0 items fetched

File: dll/shellext/zipfldr/CExplorerCommand.cpp [hunk 10]
```diff
@@ -0,0 +1,189 @@
...
+        return S_FALSE;
+    }
+    STDMETHODIMP Reset()
+    {
+        m_bFirst = true;
+        return S_OK;
+    }
+    STDMETHODIMP Clone(IEnumExplorerCommand **ppenum)
+    {
+        return E_NOTIMPL;
```
Comment (sanchaez) [inline dll/shellext/zipfldr/CExplorerCommand.cpp]: `UNIMPLEMENTED;`?

File: dll/shellext/zipfldr/CFolderViewCB.cpp [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+
+#include "precomp.h"
+
+class CFolderViewCB :
+    public CComObjectRootEx<CComMultiThreadModelNoCS>,
+    public IShellFolderViewCB
+{
+public:
+
+    CFolderViewCB()
```
Comment (sanchaez) [inline dll/shellext/zipfldr/CFolderViewCB.cpp]: This is created for you.

File: dll/shellext/zipfldr/CZipEnumerator.hpp [hunk 12]
```diff
@@ -0,0 +1,94 @@
...
+ * LICENSE:     GPL-2.0+ (https://spdx.org/licenses/GPL-2.0+)
+ * PURPOSE:     CZipEnumerator
+ * COPYRIGHT:   Copyright 2017 Mark Jansen (mark.jansen@reactos.org)
+ */
+
+struct CZipEnumerator
+{
+private:
+    CComPtr<IZip> spZip;
+    bool first;
```
Comment (sanchaez) [inline dll/shellext/zipfldr/CZipEnumerator.hpp]: `bIsFirst`

File: dll/shellext/zipfldr/CZipExtract.cpp [hunk 13]
```diff
@@ -0,0 +1,477 @@
...
+{
+    CStringW m_Filename;
+    CStringW m_Directory;
+    unzFile uf;
+public:
+    CZipExtract(PCWSTR Filename)
+        :uf(NULL)
+    {
+        m_Filename = Filename;
+        m_Directory = m_Filename;
```
Comment (sanchaez) [inline dll/shellext/zipfldr/CZipExtract.cpp]: Initializing those two members in initalizer list would be cleaner IMO.

File: dll/shellext/zipfldr/CZipExtract.cpp [hunk 14]
```diff
@@ -0,0 +1,477 @@
...
+        PWSTR Dir = m_Directory.GetBuffer();
+        PathRemoveExtensionW(Dir);
```
Comment (sanchaez) [inline dll/shellext/zipfldr/CZipExtract.cpp]: Using rvalue from GetBuffer() would be cleaner IMO.

File: dll/shellext/zipfldr/CZipExtract.cpp [hunk 15]
```diff
@@ -0,0 +1,477 @@
...
+        m_Directory.ReleaseBuffer();
+    }
+
+    ~CZipExtract()
```
Comment (sanchaez) [inline dll/shellext/zipfldr/CZipExtract.cpp]: This is created for you. Maybe you wanted virtual dtor?

File: dll/shellext/zipfldr/CZipExtract.cpp [hunk 16]
```diff
@@ -0,0 +1,477 @@
...
+    {
+    }
+
+    void Close()
+    {
+        if (uf)
+            unzClose(uf);
+        uf = NULL;
```
Comment (sanchaez) [inline dll/shellext/zipfldr/CZipExtract.cpp]: `if(uf)

File: dll/shellext/zipfldr/CZipExtract.cpp [hunk 17]
```diff
@@ -0,0 +1,477 @@
...
+        static DialogResult ShowDlg(HWND hDlg, PCSTR FullPath)
+        {
+            PCSTR Filename = PathFindFileNameA(FullPath);
+            CConfirmReplace confirm(Filename);
+            INT_PTR Result = confirm.DoModal(hDlg);
+            switch (Result)
+            {
+            case IDYES: return Yes;
+            case IDYESALL: return YesToAll;
+            default:
```
Comment (sanchaez) [inline dll/shellext/zipfldr/CZipExtract.cpp]: Seems weird but OK.

File: dll/shellext/zipfldr/CZipExtract.cpp [hunk 18]
```diff
@@ -0,0 +1,477 @@
...
+        }
+
+        LRESULT OnInitDialog(UINT nMsg, WPARAM wParam, LPARAM lParam, BOOL& bHandled)
+        {
+            CenterWindow(GetParent());
+
+            HICON hIcon = LoadIcon(NULL, IDI_EXCLAMATION);
+            SendDlgItemMessage(IDC_EXCLAMATION_ICON, STM_SETICON, (WPARAM)hIcon);
+
+            /* Our CString does not support FormatMessage yet */
```
Comment (sanchaez) [inline dll/shellext/zipfldr/CZipExtract.cpp]: Consider implementing it then 😺

File: dll/shellext/zipfldr/CZipExtract.cpp [hunk 19]
```diff
@@ -0,0 +1,477 @@
...
+                        break;
+                    }
+                    if (dwWritten != (DWORD)err)
+                    {
+                        DPRINT1("ERROR, WriteFile: dwWritten:%d err:%d\n", dwWritten, err);
+                        break;
+                    }
+                }
+
+
```
Comment (sanchaez) [inline dll/shellext/zipfldr/CZipExtract.cpp]: Double newline.

File: dll/shellext/zipfldr/zlib_contrib/MiniZip64_info.txt [hunk 20]
```diff
@@ -0,0 +1,74 @@
+MiniZip - Copyright (c) 1998-2010 - by Gilles Vollant - version 1.1 64 bits from Mathias Svensson
```
Comment (sanchaez) [inline dll/shellext/zipfldr/zlib_contrib/MiniZip64_info.txt]: I think zlib_contrib should be moved to zlib.

File: dll/win32/shell32/CDefView.cpp [hunk 21]
```diff
@@ -400,7 +402,6 @@ HRESULT WINAPI CDefView::Initialize(IShellFolder *shellFolder)
 {
     m_pSFParent = shellFolder;
     shellFolder->QueryInterface(IID_PPV_ARG(IShellFolder2, &m_pSF2Parent));
-
```
Comment (sanchaez) [inline dll/win32/shell32/CDefView.cpp]: Unrelated change.

File: modules/rostests/apitests/shell32/SHCreateFileExtractIconW.cpp [hunk 22]
```diff
@@ -119,85 +156,102 @@ START_TEST(SHCreateFileExtractIconW)
...
+                colorSizeRef = maskSizeRef = 0;
+                GetIconData(myIcon, colorDataRef, colorSizeRef, maskDataRef, maskSizeRef);
+
+                ok(colorSizeRef == colorSize, "Expected %lu, was %lu for %S(0x%lx,0x%x)\n", colorSizeRef, colorSize, cur.Name, cur.dwFlags, tryFlags[idFlags]);
+                ok(maskSizeRef == maskSize, "Expected %lu, was %lu for %S(0x%lx,0x%x)\n", maskSizeRef, maskSize, cur.Name, cur.dwFlags, tryFlags[idFlags]);
+
+                if (colorSizeRef == colorSize)
                 {
-                    ok(!memcmp(colorData, colorDataRef, colorSize), "Expected equal colorData for %S(%lx)\n", cur.Name, cur.dwFlags);
+                    /* Incase requested filetype does not match, the exe icon is not used! */
```
Comment (sanchaez) [inline modules/rostests/apitests/shell32/SHCreateFileExtractIconW.cpp]: `In case` 😄

File: modules/rostests/apitests/shell32/SHCreateFileExtractIconW.cpp [hunk 23]
```diff
@@ -101,6 +120,24 @@ START_TEST(SHCreateFileExtractIconW)
...
+    for (size_t n = 0; n < _countof(InterfaceTests); ++n)
+    {
+        {
+            CComPtr<IUnknown> spUnkown;
+            HRESULT hr = pSHCreateFileExtractIconW(L"test.txt", FILE_ATTRIBUTE_NORMAL, *InterfaceTests[n].IID, (void**)&spUnkown);
+            ok(hr == InterfaceTests[n].ExpectedCreate, "Expected hr to be 0x%lx, was 0x%lx for %u\n", InterfaceTests[n].ExpectedCreate, hr, n);
+        }
+
+        {
+            CComPtr<IUnknown> spUnkown, spUnknown2;
```
Comment (sanchaez) [inline modules/rostests/apitests/shell32/SHCreateFileExtractIconW.cpp]: `spUnknown`

File: modules/rostests/apitests/shell32/SHCreateFileExtractIconW.cpp [hunk 24]
```diff
@@ -101,6 +120,24 @@ START_TEST(SHCreateFileExtractIconW)
         return;
     }
 
+    for (size_t n = 0; n < _countof(InterfaceTests); ++n)
+    {
+        {
+            CComPtr<IUnknown> spUnkown;
```
Comment (sanchaez) [inline modules/rostests/apitests/shell32/SHCreateFileExtractIconW.cpp]: `spUnknown`

File: dll/shellext/zipfldr/CExplorerCommand.cpp [hunk 25]
```diff
@@ -0,0 +1,189 @@
...
+    {
+        if (!pUICommand)
+            return E_POINTER;
+
+        if (pceltFetched)
+            *pceltFetched = 0;
+        if (m_bFirst && celt)
+        {
+            m_bFirst = false;
+            celt--;
```
Comment (sanchaez) [inline dll/shellext/zipfldr/CExplorerCommand.cpp]: Seems redundant.

File: dll/shellext/zipfldr/zlib_contrib/MiniZip64_info.txt (hunk 20 above)
Comment (HBelusca) [reply dll/shellext/zipfldr/zlib_contrib/MiniZip64_info.txt]: You mean, inside our existing zlib?

Comment (blackcrack) [reply dll/shellext/zipfldr/zlib_contrib/MiniZip64_info.txt]: Being close to linux may not be bad. Some it is more easy if it's an bit more equivalent in direction linux as the "original", because later it's more easy to use with.. :)

File: dll/shellext/zipfldr/CZipEnumerator.hpp (hunk 5 above)
Comment (learn-more) [reply dll/shellext/zipfldr/CZipEnumerator.hpp]: I will rename those to be consistent.

File: modules/rostests/apitests/shell32/IShellFolderViewCB.cpp (hunk 3 above)
Comment (learn-more) [reply modules/rostests/apitests/shell32/IShellFolderViewCB.cpp]: I prefer apitests to be stand-alone.
When we will use this in the code, I'll move this somewhere shared.

File: modules/rostests/apitests/shell32/SHCreateFileExtractIconW.cpp (hunk 24 above)
Comment (learn-more) [reply modules/rostests/apitests/shell32/SHCreateFileExtractIconW.cpp]: nice.

File: dll/shellext/zipfldr/zlib_contrib/MiniZip64_info.txt (hunk 20 above)
Comment (ThFabba) [reply dll/shellext/zipfldr/zlib_contrib/MiniZip64_info.txt]: Since they come from the same source, keeping them together makes it easier to sync.

Comment (blackcrack) [reply dll/shellext/zipfldr/zlib_contrib/MiniZip64_info.txt]: Mark, do not mob around here otherwise I'll report you to Github.

Comment (yagoulas) [reply dll/shellext/zipfldr/zlib_contrib/MiniZip64_info.txt]: @blackcrack : I don't understand your first message either...

Comment (blackcrack) [reply dll/shellext/zipfldr/zlib_contrib/MiniZip64_info.txt]: ah this is a other musik :) Okey maybe a bit Awkwardly expressed :)
i have meant , because Linux use the zlib also because i have mentored to use the zlib too.
Makes may later usable if we use it directly.. {less complications or so.. (it was just an assumption)} ;)

File: dll/win32/shell32/CDefView.cpp [hunk 26]
```diff
@@ -402,6 +402,7 @@ HRESULT WINAPI CDefView::Initialize(IShellFolder *shellFolder)
 {
     m_pSFParent = shellFolder;
     shellFolder->QueryInterface(IID_PPV_ARG(IShellFolder2, &m_pSF2Parent));
+
```
Comment (HBelusca) [inline dll/win32/shell32/CDefView.cpp]: What does this achieve? :P

File: modules/rostests/apitests/shell32/SHCreateFileExtractIconW.cpp (hunk 22 above)
Comment (HBelusca) [reply modules/rostests/apitests/shell32/SHCreateFileExtractIconW.cpp]: `Justin Case` :laughing:

File: dll/win32/shell32/CDefView.cpp (hunk 26 above)
Comment (learn-more) [reply dll/win32/shell32/CDefView.cpp]: It negates the removal of this line in an earlier commit.

**Comment by learn-more** [comment]
Moved minizip to zlib/contrib, and into a static library.

File: sdk/lib/3rdparty/zlib/CMakeLists.txt [hunk 27]
```diff
@@ -21,9 +21,22 @@ list(APPEND SOURCE
     infback.c
     uncompr.c)
 
+list(APPEND MINIZIP_SOURCE
```
Comment (sanchaez) [inline sdk/lib/3rdparty/zlib/CMakeLists.txt]: Maybe MiniZip deserves it's own CMakeLists.txt?

File: sdk/lib/3rdparty/zlib/CMakeLists.txt [hunk 28]
```diff
@@ -21,9 +21,22 @@ list(APPEND SOURCE
...
+    contrib/minizip/unzip.c
+    contrib/minizip/unzip.h
+    contrib/minizip/zip.c
+    contrib/minizip/zip.h)
+
+
 if(CMAKE_CROSSCOMPILING)
     add_library(zlib ${SOURCE} ${SOLO_SOURCE})
     add_library(zlib_solo ${SOLO_SOURCE})
+    add_library(minizip ${MINIZIP_SOURCE})
```
Comment (learn-more) [inline sdk/lib/3rdparty/zlib/CMakeLists.txt]: missing dependency on xdk

File: sdk/lib/3rdparty/zlib/CMakeLists.txt (hunk 27 above)
Comment (learn-more) [reply sdk/lib/3rdparty/zlib/CMakeLists.txt]: Adding it inside an additional `add_subdirectory` would just be a pointless indirection.

What is there to see in the Makefile?
